qiskit/circuit/library/standard_gates/equivalence_library.py

```diff
@@ -355,14 +385,62 @@
 def_tdg.append(U1Gate(-pi / 4), [q[0]], [])
 _sel.add_equivalence(TdgGate(), def_tdg)
 
-# U2Gate
+# UGate
 
 q = QuantumRegister(1, 'q')
+theta = Parameter('theta')
 phi = Parameter('phi')
 lam = Parameter('lam')
-def_u2 = QuantumCircuit(q)
-def_u2.append(U3Gate(pi / 2, phi, lam), [q[0]], [])
-_sel.add_equivalence(U2Gate(phi, lam), def_u2)
+u_to_u3 = QuantumCircuit(q)
+u_to_u3.u3(theta, phi, lam, 0)
+_sel.add_equivalence(UGate(theta, phi, lam), u_to_u3)
+
+q = QuantumRegister(1, 'q')
+theta = Parameter('theta')
+phi = Parameter('phi')
+lam = Parameter('lam')
+u_to_rxrz = QuantumCircuit(q, global_phase=(lam + phi) / 2)
+u_to_rxrz.rz(lam, 0)
+u_to_rxrz.rx(pi/2, 0)
+u_to_rxrz.rz(theta+pi, 0)
+u_to_rxrz.rx(pi/2, 0)
+u_to_rxrz.rz(phi+3*pi, 0)
+_sel.add_equivalence(UGate(theta, phi, lam), u_to_rxrz)
+
+# CUGate
+
+q = QuantumRegister(2, 'q')
+theta = Parameter('theta')
+phi = Parameter('phi')
+lam = Parameter('lam')
+gamma = Parameter('gamma')
+def_cu = QuantumCircuit(q)
+def_cu.p(gamma, 0)
+def_cu.p((lam + phi) / 2, 0)
+def_cu.p((lam - phi) / 2, 1)
+def_cu.cx(0, 1)
+def_cu.u3(-theta / 2, 0, -(phi + lam) / 2, 1)  # TODO use QuantumCircuit.u but fails a test
```

Comment:
@kdk I don't know what's going on there, it seems like a parameter binding issue like the one we discussed before where we had a CPhase whose parameter wasn't properly bound.

Reply:
I think what's missing is a cast of params to `float` in `CUGate.to_matrix` (similar to https://github.com/Qiskit/qiskit-terra/blob/8f717d9/qiskit/circuit/library/standard_gates/u3.py#L95 ).

Reply:
that fixed it! done in dfb2af9